plugins/Utils/windowstatestorage.cpp

```diff
@@ -16,167 +17,319 @@
+        return result;
+    } else {
+        qWarning() << "getStage result expected integer, got " << resultStr;
+        return -1;
+    }
 }
 
-int WindowStateStorage::getStage(const QString &appId, int defaultValue) const
+void AsyncQuery::logSqlError(const QSqlQuery query) const
 {
-    const QString queryString = QStringLiteral("SELECT stage FROM stage WHERE appId = '%1';")
-            .arg(sanitiseString(appId));
+    qWarning() << "Error executing query" << query.lastQuery()
+               << "Driver error:" << query.lastError().driverText()
+               << "Database error:" << query.lastError().databaseText();
+}
 
-    QSqlQuery query = getValue(queryString);
+const QString AsyncQuery::getDbName()
+{
+    QSqlDatabase connection = QSqlDatabase::database(m_connectionName);
+    return connection.databaseName();
+}
 
-    if (!query.first()) {
-        return defaultValue;
+WindowStateStorage::WindowStateStorage(const QString& dbName, QObject *parent):
+    QObject(parent),
+    m_thread()
+{
+    QString dbFile;
+    if (dbName == nullptr) {
+        const QString dbPath = QStandardPaths::writableLocation(QStandardPaths::GenericCacheLocation) + QStringLiteral("/unity8/");
+        QDir dir;
+        dir.mkpath(dbPath);
+        dbFile = QString(dbPath) + QStringLiteral("windowstatestorage.sqlite");
+    } else {
+        dbFile = dbName;
+    }
+    m_asyncQuery = new AsyncQuery(dbFile);
+    m_asyncQuery->moveToThread(&m_thread);
+    connect(&m_thread, &QThread::finished, m_asyncQuery, &QObject::deleteLater);
+    m_thread.start();
+    bool queryInitSuccessful;
+    QMetaObject::invokeMethod(m_asyncQuery, "initdb",
+                              Qt::BlockingQueuedConnection,
+                              Q_RETURN_ARG(bool, queryInitSuccessful));
+    if (!queryInitSuccessful) {
+        qWarning() << "Failed to initialize WindowStateStorage! Windows will not be restored to their previous location.";
+        m_asyncOk = false;
+        m_thread.quit();
+    } else {
+        connect(this, &WindowStateStorage::saveGeometry, m_asyncQuery, &AsyncQuery::saveGeometry);
+        connect(this, &WindowStateStorage::saveStage, m_asyncQuery, &AsyncQuery::saveStage);
+        m_asyncOk = true;
     }
-    return query.value("stage").toInt();
 }
 
-QRect WindowStateStorage::getGeometry(const QString &windowId, const QRect &defaultValue) const
+WindowStateStorage::~WindowStateStorage()
 {
-    QString queryString = QStringLiteral("SELECT * FROM geometry WHERE windowId = '%1';")
-            .arg(sanitiseString(windowId));
+    m_thread.quit();
+    m_thread.wait();
+}
 
-    QSqlQuery query = getValue(queryString);
+void WindowStateStorage::saveState(const QString &windowId, WindowStateStorage::WindowState windowState)
+{
+    if (!m_asyncOk) {
+        return;
+    }
+    // This could be a simple connection like all the other save* methods
+    // on AsyncQuery, but WindowStateStorage::WindowState can't be used
+    // on it directly.
```

Comment:
I think there should be a way to use WindowStateStorage::WindowState as an argument to AsyncQuery::saveState, but I haven't been able to find it yet. It compiles if I flip the order of AsyncQuery and WindowStateStorage in the header file and forward-declare AsyncQuery, but the connect call fails saying that WindowStateStorage::WindowState is not a registered metatype.

Reply:
I got it, but I'm not sure if I did it the best way possible. Where should I put the qRegisterMetaType call? https://github.com/ubports/unity8/pull/356/commits/02ca0715eae67635a597bd1ced218107a5b21e86